### ISSUE TITLE
Istio ingress-gateway support for health check

### ISSUE THREAD

**Issue by quanjielin** [istio/istio#9385]
**Describe the feature request**
In some cases, Istio ingress-gateway acts as 2nd layer load balancer(it becomes backend service of 1st layer LB). Without istio ingress-gateway support for health check, 1st layer LB cannot tell backend service status, which result in 1st layer LB in failed status as well.   

**Possible solution**
add [envoy http health check filter](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/health_checking) in istio ingress-gateway

**Comment by quanjielin** [comment]
From https://github.com/istio/istio/blob/master/pilot/pkg/networking/plugin/health/health.go#L99, istio already has health plugin but only enable for sidecar now, need the change to enable it for gateway.

**Comment by draeron** [comment]
hi there, is there any workaround for istio 1.0.3? right now everytimes we have a new deployment we're having issues where the gateway sends requests to the deads pods IPs.

**Comment by cwocwo** [comment]
Istio ingress-gateway should support for health check.

We have two ingressgateway pods, for some reason we restart the two pods, unfortunately one of the pods made an error, but it's `status` was `Running`.    This cause the restful api through virtualservice show errors sometimes.

### istio pods status
```
root@k8s01:~# kubectl -n istio-system get po
NAME                                      READY     STATUS    RESTARTS   AGE
grafana-9cfc9d4c9-vsw9d                   1/1       Running   0          10d
istio-citadel-688c4c674c-5k4n2            1/1       Running   0          25s
istio-egressgateway-76c8656b68-fh2gs      1/1       Running   0          10d
istio-galley-58589475dd-jrmwj             1/1       Running   0          29d
istio-ingressgateway-5c7bc9d579-8lkt2     1/1       Running   0          1m
istio-ingressgateway-5c7bc9d579-hhbvm     1/1       Running   0          27m
istio-pilot-57584f64b4-f8g64              1/2       Running   0          25s
istio-pilot-57584f64b4-q8fhf              1/2       Running   0          25s
istio-policy-657b668bd-rlrzq              2/2       Running   0          25s
istio-policy-657b668bd-xlbjn              2/2       Running   0          25s
istio-sidecar-injector-596bcddbb6-fpx9l   1/1       Running   0          8d
istio-telemetry-6f996864b9-6kg62          2/2       Running   3          29d
istio-telemetry-6f996864b9-9vbs5          2/2       Running   0          10d
istio-tracing-7956d69794-pczgr            1/1       Running   0          10d
```


### The error pod's log
```
kubectl -n istio-system logs istio-ingressgateway-5c7bc9d579-8lkt2 -f 
```

```
Flag shorthand -v has been deprecated, please use --log_output_level instead
Flag --v has been deprecated, please use --log_output_level instead
2019-01-10T02:44:21.295235Z	info	Version root@8c2feba0b568-docker.io/istio-1.0.4-d5cb99f479ad9da88eebb8bb3637b17c323bc50b-Clean
2019-01-10T02:44:21.295284Z	info	Proxy role: model.Proxy{ClusterID:"", Type:"router", IPAddress:"172.31.218.4", ID:"istio-ingressgateway-5c7bc9d579-2p299.istio-system", Domain:"istio-system.svc.cluster.local", Metadata:map[string]string(nil)}
2019-01-10T02:44:21.295573Z	info	Effective config: binaryPath: /usr/local/bin/envoy
configPath: /etc/istio/proxy
connectTimeout: 10s
discoveryAddress: istio-pilot:8080
discoveryRefreshDelay: 1s
drainDuration: 45s
parentShutdownDuration: 60s
proxyAdminPort: 15000
serviceCluster: istio-ingressgateway
zipkinAddress: zipkin:9411

2019-01-10T02:44:21.295590Z	info	Monitored certs: []envoy.CertSource{envoy.CertSource{Directory:"/etc/certs/", Files:[]string{"cert-chain.pem", "key.pem", "root-cert.pem"}}}
2019-01-10T02:44:21.295722Z	info	Starting proxy agent
2019-01-10T02:44:21.296838Z	info	Received new config, resetting budget
2019-01-10T02:44:21.297113Z	info	Reconciling configuration (budget 10)
2019-01-10T02:44:21.297142Z	info	Epoch 0 starting
2019-01-10T02:44:21.297782Z	info	Envoy command: [-c /etc/istio/proxy/envoy-rev0.json --restart-epoch 0 --drain-time-s 45 --parent-shutdown-time-s 60 --service-cluster istio-ingressgateway --service-node router~172.31.218.4~istio-ingressgateway-5c7bc9d579-2p299.istio-system~istio-system.svc.cluster.local --max-obj-name-len 189 --allow-unknown-fields -l warn --v2-config-only]
[2019-01-10 02:44:21.313][20][info][main] external/envoy/source/server/server.cc:190] initializing epoch 0 (hot restart version=10.200.16384.256.options=capacity=16384, num_slots=8209 hash=228984379728933363 size=4882536)
[2019-01-10 02:44:21.313][20][info][main] external/envoy/source/server/server.cc:192] statically linked extensions:
[2019-01-10 02:44:21.313][20][info][main] external/envoy/source/server/server.cc:194]   access_loggers: envoy.file_access_log,envoy.http_grpc_access_log
[2019-01-10 02:44:21.313][20][info][main] external/envoy/source/server/server.cc:197]   filters.http: envoy.buffer,envoy.cors,envoy.ext_authz,envoy.fault,envoy.filters.http.header_to_metadata,envoy.filters.http.jwt_authn,envoy.filters.http.rbac,envoy.grpc_http1_bridge,envoy.grpc_json_transcoder,envoy.grpc_web,envoy.gzip,envoy.health_check,envoy.http_dynamo_filter,envoy.ip_tagging,envoy.lua,envoy.rate_limit,envoy.router,envoy.squash,istio_authn,jwt-auth,mixer
[2019-01-10 02:44:21.313][20][info][main] external/envoy/source/server/server.cc:200]   filters.listener: envoy.listener.original_dst,envoy.listener.proxy_protocol,envoy.listener.tls_inspector
[2019-01-10 02:44:21.313][20][info][main] external/envoy/source/server/server.cc:203]   filters.network: envoy.client_ssl_auth,envoy.echo,envoy.ext_authz,envoy.filters.network.rbac,envoy.filters.network.thrift_proxy,envoy.http_connection_manager,envoy.mongo_proxy,envoy.ratelimit,envoy.redis_proxy,envoy.tcp_proxy,mixer
[2019-01-10 02:44:21.313][20][info][main] external/envoy/source/server/server.cc:205]   stat_sinks: envoy.dog_statsd,envoy.metrics_service,envoy.stat_sinks.hystrix,envoy.statsd
[2019-01-10 02:44:21.313][20][info][main] external/envoy/source/server/server.cc:207]   tracers: envoy.dynamic.ot,envoy.lightstep,envoy.zipkin
[2019-01-10 02:44:21.313][20][info][main] external/envoy/source/server/server.cc:210]   transport_sockets.downstream: alts,envoy.transport_sockets.capture,raw_buffer,tls
[2019-01-10 02:44:21.313][20][info][main] external/envoy/source/server/server.cc:213]   transport_sockets.upstream: alts,envoy.transport_sockets.capture,raw_buffer,tls
[2019-01-10 02:44:21.317][20][info][config] external/envoy/source/server/configuration_impl.cc:50] loading 0 static secret(s)
[2019-01-10 02:44:21.321][20][warning][upstream] external/envoy/source/common/config/grpc_mux_impl.cc:240] gRPC config stream closed: 14, no healthy upstream
[2019-01-10 02:44:21.321][20][warning][upstream] external/envoy/source/common/config/grpc_mux_impl.cc:41] Unable to establish new stream
[2019-01-10 02:44:21.321][20][info][config] external/envoy/source/server/configuration_impl.cc:60] loading 1 listener(s)
[2019-01-10 02:44:21.322][20][info][config] external/envoy/source/server/configuration_impl.cc:94] loading tracing configuration
[2019-01-10 02:44:21.322][20][info][config] external/envoy/source/server/configuration_impl.cc:103]   loading tracing driver: envoy.zipkin
[2019-01-10 02:44:21.322][20][info][config] external/envoy/source/server/configuration_impl.cc:116] loading stats sink configuration
[2019-01-10 02:44:21.323][20][info][main] external/envoy/source/server/server.cc:432] starting main dispatch loop
[2019-01-10 02:44:21.324][20][info][upstream] external/envoy/source/common/upstream/cluster_manager_impl.cc:130] cm init: initializing cds
```

**Comment by ChrisCooney** [comment]
Any thoughts on this from the core contributors?

**Comment by mandarjog** [comment]
Readiness for gateways is available in 1.1-RC0.
```
2019-02-20T23:03:57.309092Z     info    Version root@9efe8d69-34c2-11e9-b805-0a580a2c0506-docker.io/istio-1.1.0-rc.0-3632361b439ae27962f0af472064d452fc63273b-Clean
2019-02-20T23:03:57.309390Z     info    Obtained private IP [10.16.0.144 fe80::b8ba:81ff:fe4c:122]

2019-02-20T23:03:57.309462Z     info    Proxy role: &model.Proxy{ClusterID:"", Type:"router", IPAddresses:[]string{"10.16.0.144", "10.16.0.144", "fe80::b8ba:81ff:fe4c:122"}, ID:"istio-ingressgatew
ay-57c8948b64-7l46x.istio-system", Locality:(*core.Locality)(nil), DNSDomains:[]string(nil), ConfigNamespace:"", TrustDomain:"cluster.local", Metadata:map[string]string(nil), SidecarScope:(*model.
SidecarScope)(nil), ServiceInstances:[]*model.ServiceInstance(nil)}
2019-02-20T23:03:57.309495Z     info    PilotSAN []string{"spiffe://cluster.local/ns/istio-system/sa/istio-pilot-service-account"}
2019-02-20T23:03:57.309953Z     info    Effective config: binaryPath: /usr/local/bin/envoy
configPath: /etc/istio/proxy
connectTimeout: 10s
controlPlaneAuthPolicy: MUTUAL_TLS
discoveryAddress: istio-pilot:15011
drainDuration: 45s
parentShutdownDuration: 60s
proxyAdminPort: 15000
serviceCluster: istio-ingressgateway
statNameLength: 189
tracing:
  zipkin:
    address: zipkin:9411


2019-02-20T23:05:12.237210Z     info    Envoy proxy is NOT ready: cds updates: 0 successful, 0 rejected; lds updates: 0 successful, 0 rejected
2019-02-20T23:05:14.237405Z     info    Envoy proxy is NOT ready: cds updates: 0 successful, 0 rejected; lds updates: 0 successful, 0 rejected
2019-02-20T23:05:16.237399Z     info    Envoy proxy is ready
```

**Comment by vadimi** [comment]
@ChrisCooney it's actually available already in istio 1.0.6. We use `/healthz/ready` endpoint with AWS ALB. The only problem is it is disabled by default, you need to enable it by using
```
          - --statusPort
          - "15020"
```
in istio-ingressgateway. Then you need to map port 15020 to whatever NodePort you need to feed it to the loadbalancer config.

**Comment by ChrisCooney** [comment]
> @ChrisCooney it's actually available already in istio 1.0.6. We use `/healthz/ready` endpoint with AWS ALB. The only problem is it is disabled by default, you need to enable it by using
> 
> ```
>           - --statusPort
>           - "15020"
> ```
> in istio-ingressgateway. Then you need to map port 15020 to whatever NodePort you need to feed it to the loadbalancer config.

Awesome! Thanks @vadimi , i'll take a look. Right now we've got Traefik -> ingress gateway, so could be nice to remove the extra moving part.
